### DISCUSSION TITLE
Auto-Generate Property Documentation

### DISCUSSION THREAD

**Discussion by robloo** [AvaloniaUI/Avalonia#11198] [General]
I'm putting this in the main Avalonia repository for more visibility and because there are two different documentation repos right now and both don't really use discussions.

---

In writing the ColorPicker docs I found I was going to have to copy-paste a lot of the source code comments for styled properties and properties themselves. This is wasting time and will quickly become outdated/desynced. My suggestion is to have a script that can fill in properties for a specified type (optionally supporting inheritance, etc.).

Then inside the .md file you could use a special syntax telling the script where to update the properties.

```md
[//]: # (BeginPropertySection<ColorView>)

...

[//]: # (EndPropertySection)
```

This is using some syntax tricks (https://stackoverflow.com/questions/4823468/comments-in-markdown) to insert comments in the markdown. The comments would then be picked up by the script and everything between the begin/end tags would be deleted and filled in with generated property docs.

Thoughts?

**Comment by maxkatz6** [comment]
I wish partial properties were introduced in C#12. https://github.com/dotnet/csharplang/issues/6420
This way we could define only a CLR property with an attribute. And everything else including getter, setter and static definition with comments could be autogenerated. 

But it seems it's not out for yet another year.

**Reply by robloo** [reply]
Yea... I've been following partial properties for a while too. Everyone is waiting for it so I'm surprised it was de-prioritized. MVVM toolkit is waiting too.

I think the partial property itself would still need comments though (to be copied over or shared with the generated implementation). Comments can't be auto-generated. Unless, you are saying we reverse the process and document properties in the docs themselves and then just include that text in the auto-generated properties?

Either way, I think the syntax above for the markdown text target for inserting generated values should be considered.

**Reply by maxkatz6** [reply]
> Comments can't be auto-generated.

Why not?
We still need to write documentation to the CLR properties on which we add attributes.
But styled/direct property definitions usually have a pretty typical repeating documentation, which can be generated based on the property name only.

And same attribute can be later used to detect which properties need to be included in the Markdown documentation.

**Reply by maxkatz6** [reply]
Specifically, about documentation website, previously it was autogenerated by some tool that @grokys might know. And used on the reference page https://reference.avaloniaui.net/api/Avalonia.Controls/Button/

More user-friendly doc pages have a more limited selection of API members https://docs.avaloniaui.net/docs/controls/buttons/button

**Reply by maxkatz6** [reply]
CC @MikeCodesDotNET

**Reply by robloo** [reply]
Yes, Property definition comments can absolutely be generated. I was talking about the CLR property comments which require some thought and maybe even useful remarks. Those are the comments that need to be synced with the docs anyway.

> And same attribute can be later used to detect which properties need to be included in the Markdown documentation.

Microsoft auto-generates the docs as well I think. Perhaps an attribute to ignore from docs is useful but I wouldn't suggest any to include. Everything in the type should be included by default. Microsofts standard of not including inherited properties in derived types is important as well.
